### PR TITLE
test: Increase unit test code coverage in pkg/shared

### PR DESCRIPTION
**Pull request by VanillaSpoon**
## Description
Increase unit test code coverage in pkg/shared

[MGDSTRM-7989](https://issues.redhat.com/browse/MGDSTRM-7989)

filename| starting coverage % | finishing coverage %|
-- | -- | --
stringscanner/simple_string_scanner.go | 81 | 95.2
stringscanner/sql_string_scanner.go | 77.4 | 83.9
util/arrays/generic_array_util.go | 75 | 100
diff.go | 0 | 70
util.go | 0 | 100
load_openapi_spec.go | 0 | 100
util/statemachine/grammar.go | 0 | 100
util/statemachine/statemachine.go | 92.9 |  
numutil.go | 100 |  
util/arrays/string_array_util.go | 100 |  
get_project_root_dir.go | 93.3 |  
config.go | 82.9 |  
util/statemachine/builder.go | 77.8 |  
secrets/secrets.go 69.8 | 69.8 |  

filename| starting coverage % | finishing coverage %|
-- | -- | --
handle_error.go | 0 | add to ignore
write_json_response.go | 0 | add to ignore

is_jwt_token_expired.go --> Will be moved to pkg/auth in a follow up pr

## Checklist (Definition of Done)

- [x] All acceptance criteria specified in JIRA have been completed
- [x] Unit and integration tests added that prove the fix is effective or the feature works (tested against emulated and non-emulated OCM environment)
- [ ] ~~Documentation added for the feature~~
- [x] CI and all relevant tests are passing
- [x] Code Review completed
- [x] Verified independently by reviewer
- [ ] ~~Required metrics/dashboards/alerts have been added (or PR created).~~
- [ ] ~~Required Standard Operating Procedure (SOP) is added.~~
- [ ] ~~JIRA has created for changes required on the client side~~

### CODE REVIEW

**Comment by VanillaSpoon** [comment]
Hey @machi1990 @pawelpaszki  would either of you disagree with the files I have marked as planning to ignore in the coverage report?

**Comment by machi1990** [comment]
> would either of you disagree with the files I have marked as planning to ignore in the coverage report?

Thanks @VanillaSpoon for tagging me along the PR. Below is my response,

> handle_error.go
> write_json_response.go

yes for these two, I think they are best tested at the integration level.

> is_jwt_token_expired.go
> load_openapi_spec.go
> util/statemachine/grammar.go

I think these three can be unit tested.

What do you think @pawelpaszki @VanillaSpoon ?

**Comment by VanillaSpoon** [comment]
Thanks @machi1990

File: pkg/shared/utils/arrays/generic_array_utils_test.go [hunk 1]
```diff
@@ -1,33 +1,75 @@
...
-			return val > 5
-		} else {
-			return false
-		}
-	}, 1, 2, "Hello", 4, 5, 6, 7, 8)
-	Expect(idx).To(Equal(5))
-	Expect(val).To(Equal(6))
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, got1 := FindFirst(tt.args.predicate, tt.args.values...)
```
Comment (Rajagopalan-Ranganathan) [inline pkg/shared/utils/arrays/generic_array_utils_test.go:68]: got, got1 are not good variable names may be receivedIndex, receivedVal or gotIndex, gotVal might be better

File: pkg/shared/utils/stringscanner/simple_string_scanner_test.go [hunk 2]
```diff
@@ -48,3 +49,62 @@ func Test_SimpleScanner(t *testing.T) {
...
+
+	RegisterTestingT(t)
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			s := &simpleStringScanner{
+				value: tt.fields.value,
+				pos:   tt.fields.pos,
+			}
+			got, got1 := s.Peek()
```
Comment (Rajagopalan-Ranganathan) [inline pkg/shared/utils/stringscanner/simple_string_scanner_test.go:105]: variable names - let them be self explanatory.

File: pkg/shared/utils/stringscanner/sql_string_scanner_test.go [hunk 3]
```diff
@@ -80,3 +81,70 @@ func Test_SQLStringScanner(t *testing.T) {
...
+
+	RegisterTestingT(t)
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			s := &scanner{
+				tokens: tt.fields.tokens,
+				pos:    tt.fields.pos,
+			}
+			got, got1 := s.Peek()
```
Comment (Rajagopalan-Ranganathan) [inline pkg/shared/utils/stringscanner/sql_string_scanner_test.go:145]: variable names - let them be self explanatory.

Comment (VanillaSpoon) [reply pkg/shared/utils/stringscanner/sql_string_scanner_test.go]: Thanks for the feedback @Rajagopalan-Ranganathan

File: pkg/shared/diff_test.go [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+			name: "return JSON of diff from arguments",
+			args: args{
+				a:     "word1",
+				b:     "word2",
+				aName: "testNameA",
+				bName: "testNameB",
+			},
+			want: "\n--- testNameA\n+++ testNameB\n@@ -1 +1 @@\n-\"word1\"\n+\"word2\"\n",
+		},
+	}
```
[the remark this answers is not in the document]
Comment (VanillaSpoon) [reply pkg/shared/diff_test.go:81]: Thanks for the feedback, I've applied the additions @ziccardi

**Comment by pawelpaszki** [comment]
> > would either of you disagree with the files I have marked as planning to ignore in the coverage report?
> 
> Thanks @VanillaSpoon for tagging me along the PR. Below is my response,
> 
> > handle_error.go
> > write_json_response.go
> 
> yes for these two, I think they are best tested at the integration level.
> 
> > is_jwt_token_expired.go
> > load_openapi_spec.go
> > util/statemachine/grammar.go
> 
> I think these three can be unit tested.
> 
> What do you think @pawelpaszki @VanillaSpoon ?

sorry. only saw this comment now. I agree with Manyanda

**Review by valdar — APPROVED** [review]

File: pkg/shared/load_openapi_spec_test.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+		},
+	}
+
+	RegisterTestingT(t)
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := LoadOpenAPISpec(tt.args.assetFunc, tt.args.asset)
+			Expect(got).To(Equal(tt.want))
+			Expect(err != nil).To(Equal(tt.wantErr))
```
Comment (ziccardi) [inline pkg/shared/load_openapi_spec_test.go:59]: If an unexpected error occurres, `got` will be nil and the `Expect(got).To...` will fail, halting the test execution.
That means that the `Expect(err != nil)` won't be executed, resulting in confusing errors (error will fail for unmatched `got` instead than unexpected error).

I'd suggest to always move the `Expect(err...)` statements as first check

File: pkg/shared/util_test.go [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+	}{
+		{
+			name: "return 0 if ptr is nil",
+			args: args{
+				ptr: nil,
+			},
+			want: 0,
+		},
+		{
+			name: "return string if ptr is not nil",
```
Comment (ziccardi) [inline pkg/shared/util_test.go:63]:
```suggestion
			name: "return int64 if ptr is not nil",
```

File: pkg/shared/utils/arrays/generic_array_utils_test.go [hunk 7]
```diff
@@ -1,33 +1,75 @@
...
-	}, 1, 2, "Hello", 4, 5, 6, 7, 8)
-	Expect(idx).To(Equal(5))
-	Expect(val).To(Equal(6))
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			gotIndex, gotVal := FindFirst(tt.args.predicate, tt.args.values...)
+			Expect(gotIndex).To(Equal(tt.want))
+			if gotVal != nil {
+				Expect(gotVal).To(Equal(tt.want1))
+			}
```
Comment (ziccardi) [inline pkg/shared/utils/arrays/generic_array_utils_test.go:72]: What if `gotVal` is `nil` but a value was expected? With this test, the case is skipped.
I suggest the change below:
```suggestion
	Expect(gotVal).To(Equal(tt.want1))
```

**Review by ziccardi — CHANGES_REQUESTED** [review]
Left some comment

**Review by ziccardi — COMMENTED** [review]
lgtm

File: pkg/shared/utils/arrays/generic_array_utils_test.go [hunk 8]
```diff
@@ -1,33 +1,75 @@
...
+	. "github.com/onsi/gomega"
 )
 
-func TestGenericFindFirstWithStrings(t *testing.T) {
-	RegisterTestingT(t)
-	idx, val := FindFirst(func(x interface{}) bool { return x.(string) == "Red" }, "This", "Is", "Red", "Hat")
-	Expect(idx).To(Equal(2))
-	Expect(val).To(Equal("Red"))
-}
+func TestFindFirst(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (JameelB) [reply pkg/shared/utils/arrays/generic_array_utils_test.go]: Is this a new naming convention for all of our tests (unit and integration)?

File: pkg/shared/diff_test.go [hunk 9]
```diff
@@ -0,0 +1,91 @@
+package shared
+
+import (
+	"testing"
+
+	. "github.com/onsi/gomega"
+)
+
+func Test_DiffAsJson(t *testing.T) {
```
Comment (JameelB) [inline pkg/shared/diff_test.go:9]: Is this function still used? I don't see any reference of it being called.

Comment (VanillaSpoon) [reply pkg/shared/diff_test.go:9]: Great question @JameelB 
The method was added [here](https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/commit/c388577e4db0b1fb49dcffb61cbc3ec338501743)\
But I haven't spotted any uses of it

Comment (JameelB) [reply pkg/shared/diff_test.go:9]: Is this method safe to remove @chirino?

**Comment by JameelB** [comment]
Changes looks good to me. One pending question about that DiffAsJson but we can remove that on a follow on PR. 

@ziccardi I see you've added lgtm here but there's still 1 change requested. Is this good to merge?

**Review by ziccardi — APPROVED** [review]

**Pull request merged**